scipy/interpolate/_cubic.py

```diff
@@ -491,34 +583,177 @@ def __init__(self, x, y, axis=0, extrapolate=True):
+                    b[-1] = bc_end[1]
+                elif bc_end[0] == 2:
+                    A[1, -1] = 2 * dx[-1]
+                    A[-1, -2] = dx[-1]
+                    b[-1] = 0.5 * bc_end[1] * dx[-1]**2 + 3 * (y[-1] - y[-2])
+
+                s = solve_banded((1, 1), A, b, overwrite_ab=True,
+                                 overwrite_b=True, check_finite=False)
 
         # Compute coefficients in PPoly form.
-        c0 = (s[:-1] + s[1:] - 2 * slope) / dxr**2
-        c = np.empty((4, n - 1) + y.shape[1:], dtype=c0.dtype)
-        c[0] = c0
-        c[1] = (slope - s[:-1]) / dxr - c0 * dxr
+        t = (s[:-1] + s[1:] - 2 * slope) / dxr
+        c = np.empty((4, n - 1) + y.shape[1:], dtype=t.dtype)
+        c[0] = t / dxr
+        c[1] = (slope - s[:-1]) / dxr - t
         c[2] = s[:-1]
         c[3] = y[:-1]
 
         super(CubicSpline, self).__init__(c, x, extrapolate=extrapolate)
         self.axis = axis
+
+    @staticmethod
+    def _validate_bc(bc_type, y, expected_deriv_shape, axis):
+        """Validate and prepare boundary conditions.
+
+        Returns
+        -------
+        validated_bc : 2-tuple
+            Boundary conditions for a curve start and end.
+        """
+        if isinstance(bc_type, string_types):
+            if bc_type == 'periodic':
+                if not np.allclose(y[0], y[-1], rtol=1e-15, atol=1e-15):
+                    raise ValueError(
+                        "The first and last `y` point along axis {} must "
+                        "be identical (within machine precision) when "
+                        "bc_type='periodic'.".format(axis))
+
+            bc_type = (bc_type, bc_type)
+
+        else:
+            if len(bc_type) != 2:
+                raise ValueError("`bc_type` must contain 2 elements to "
+                                 "specify start and end conditions.")
+
+            if 'periodic' in bc_type:
+                raise ValueError("'periodic' `bc_type` is defined for both "
+                                 "curve ends and cannot be used with other "
+                                 "boundary conditions.")
+
+        validated_bc = []
+        for bc in bc_type:
+            if isinstance(bc, string_types):
+                if bc == 'clamped':
+                    validated_bc.append((1, np.zeros(expected_deriv_shape)))
+                elif bc == 'natural':
+                    validated_bc.append((2, np.zeros(expected_deriv_shape)))
+                elif bc in ['not-a-knot', 'periodic']:
+                    validated_bc.append(bc)
+                else:
+                    raise ValueError("bc_type={} is not allowed.".format(bc))
+            else:
+                try:
+                    deriv_order, deriv_value = bc
+                except Exception:
+                    raise ValueError("A specified derivative value must be "
+                                     "given in the form (order, value).")
+
+                deriv_order = int(deriv_order)
```

Comment:
This is a bit too flexible, too many things can be converted to an integer value of one:

```
In [15]: x = np.arange(8)

In [16]: s1 = CubicSpline(x, x**3, bc_type=("natural", (1.5, 0)))

In [17]: s2 = CubicSpline(x, x**3, bc_type=("natural", (True, 0)))
```